Fix #2407: Cap nproc_per_node based on CPU resources for PyTorch TrainJob

Diasker opened the pull request:
# Fix #2407: Cap nproc_per_node based on CPU resources for PyTorch TrainJob

## Problem Description

When `nproc_per_node` is set to "auto" in a PyTorch TrainJob, PyTorch's default behavior is to set it to the number of available CPUs on the node (via `os.cpu_count()`). This can cause problems when:

1. The container has a CPU limit that is lower than the node's actual CPU count
2. No GPU is requested, so PyTorch tries to use all available CPUs
3. The training job exceeds its memory limit due to too many processes, leading to OOM errors or deadlocks

This issue was discussed in PR #2387 and formally documented in Issue #2407.

## Solution

This PR modifies the Torch plugin to cap `nproc_per_node` based on CPU resources when it's set to "auto" and no GPU is requested:

1. First, check if GPU resources are requested (by examining both Limits and Requests in ResourcesPerNode)
2. If no GPU is requested, set `nproc_per_node` to the CPU limit (or request if no limit is set)
3. Default to 1 if no CPU resources are specified
4. Ensure at least 1 process, even if CPU limit is less than 1

This ensures that PyTorch distributed training only uses the CPU resources allocated to the container, rather than attempting to use all available CPUs on the node.

## Code Changes

Modified the following files:

1. `pkg/runtime/framework/plugins/torch/torch.go`: Added logic to cap `nproc_per_node` based on CPU resources
2. `pkg/runtime/framework/plugins/torch/torch_nproc_per_node_test.go`: Added test cases to validate the fix

## Test Cases

Added comprehensive test cases to verify the behavior:
- When `nproc_per_node=auto` with a CPU limit, it should be capped to that limit
- When `nproc_per_node=auto` with no CPU resources specified, it should default to 1
- When `nproc_per_node=auto` with a low CPU limit, it should be capped to that limit even if the actual CPU count is higher
- When `nproc_per_node=auto` with only CPU requests (no limits), it should use the request value
- When `nproc_per_node=auto` with GPU resources, it should remain "auto"
- When `nproc_per_node` is explicitly set, that value should be preserved

## Related Links

- Issue #2407: Cap nproc_per_node based on the CPU resources of the node for PyTorch TrainJob
- PR #2387: KEP-2170: Add PyTorch DDP MNIST training example (related discussion)

Diasker commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +74,46 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+			}
+
+			// Ensure at least 1 process
+			if cpuLimit < 1 {
```
Thank you for your valuable suggestions! I've implemented the changes you recommended:

1. Now using `resource.MustParse` to handle CPU resource values more robustly
2. Implemented `RoundUp` method to properly handle millicore cases as you suggested

The updated code now correctly handles various CPU resource formats:
- Integer values (e.g., "4")
- Decimal values (e.g., "2.5")
- Millicore values (e.g., "2500m")

I've also added comprehensive test cases to verify these scenarios, including specific tests for millicore and fractional CPU limits. All tests are passing, confirming that the implementation correctly rounds up CPU values when determining the number of processes.

This implementation ensures that when `nproc_per_node` is set to "auto" and no GPU is requested, we'll properly cap the number of processes based on the CPU resources, even when those resources are specified in millicore format.

Thanks again for helping improve the robustness of this code!

astefanutti commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +75,53 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+			cpuValue := cpuQuantity.Value()
+
+			numProcPerNode = intstr.FromInt(int(cpuValue))
+			fmt.Printf("CPU-only device detected with nproc_per_node=auto, capping to CPU limit: %d\n", cpuValue)
```
We may want to use the logger from the context, e.g. `log := ctrl.LoggerFrom(ctx)` but the context should be added first to `EnforceMLPolicy` similarly to other plugin APIs like `ComponentBuilderPlugin.Build` or `TerminalConditionPlugin.TerminalCondition`.

Diasker replied:
Thank you for your suggestion! I noticed that using `ctrl.LoggerFrom(ctx)` would require a significant change to the `EnforceMLPolicyPlugin` interface, as it currently doesn't accept a context parameter.

This change would affect:
1. The interface definition itself
2. All plugins implementing this interface (Torch, MPI, PlainML)
3. The framework code that calls these plugins
4. All related test code

Considering that the scope of this change might exceed the goals of the current PR, I'd like to ask:

1. How important is this log message? Could we temporarily keep fmt.Printf or use a logging library that doesn't require context (like klog)?
2. Or do you think we should proceed with this interface change in the current PR?

Please let me know your thoughts. Thank you!

astefanutti replied:
Yes, I agree. I'd be inclined to remove this log message for now.

@andreyvelich @tenzen-y if you have any suggestions.

tenzen-y replied:
Actually, we can add logger without interface changes in the following:

```go
type Torch struct{
	logger logr.Logger
}

func New(context.Context, client.Client, client.FieldIndexer) (framework.Plugin, error) {
	return &Torch{
		logger: ctrl.LoggerFrom(ctx).WithValues("pluginName", Name)
	}, nil
}
```

I was planning to add logger like this when I designed this interface.

astefanutti replied:
I think the main difference with both approaches is with the context from the signature it's the child context that holds a reference to the reconciled object, while if that the context from the plugin, it's the parent context.

Diasker replied:
> I think the main difference with both approaches is with the context from the signature it's the child context that holds a reference to the reconciled object, while if that the context from the plugin, it's the parent context.

Thank you for pointing out the difference between the two approaches. I initially implemented the solution using the plugin initialization context as suggested by @tenzen-y, without seeing your latest comment. Sorry about that.
I understand your point about the context differences - using the context from the method signature would provide more specific information about the reconciled object, while the plugin initialization context is more general. So, do we still need to keep or remove the log message? Actually, I think the necessity of this log message is limited.

astefanutti replied:
@Diasker thanks I agree, let's remove the message to stick to the scope of this PR as you suggested, and re-consider this when the need for more important messages in plugins like this will arise.

Diasker replied:
> @Diasker thanks I agree, let's remove the message to stick to the scope of this PR as you suggested, and re-consider this when the need for more important messages in plugins like this will arise.

@astefanutti Thanks for your guidance. I just removed the log message.

astefanutti commented:
@Diasker right I think that makes sense to update the test as you're suggesting.

Both option would work, but I'd be inclined to go with 2) change the expected value to "1" to match the CPU-based calculation.

Diasker commented:
@astefanutti Alright, change finished, thanks!

andreyvelich commented:
/ok-to-test

andreyvelich commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +75,52 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+	// Cap nproc_per_node based on CPU resources when set to "auto" and no GPU is requested
+	if numProcPerNode.String() == "auto" && trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.ResourcesPerNode != nil {
+		// Check if GPU is requested
+		hasGPU := false
+		for resourceName := range trainJob.Spec.Trainer.ResourcesPerNode.Limits {
+			if strings.Contains(strings.ToLower(string(resourceName)), "gpu") {
+				hasGPU = true
+				break
+			}
+		}
+		if !hasGPU {
+			for resourceName := range trainJob.Spec.Trainer.ResourcesPerNode.Requests {
+				if strings.Contains(strings.ToLower(string(resourceName)), "gpu") {
+					hasGPU = true
+					break
+				}
+			}
+		}
+
+		// If no GPU is requested, use CPU limit or default to 1
+		if !hasGPU {
```
@tenzen-y @astefanutti Is there any helper function in `k8s.io/utils` we can use to check whether resource (e.g. `cpu`, `gpu`) exists in container.resources ?

tenzen-y replied:
AFAIK, there is not such util.

andreyvelich commented on pkg/runtime/framework/plugins/torch/torch_nproc_per_node_test.go:
```diff
@@ -0,0 +1,265 @@
+	utiltesting "github.com/kubeflow/trainer/pkg/util/testing"
+)
+
+func TestTorch_EnforceMLPolicy_NumProcPerNode_Issue2407(t *testing.T) {
```
I think, these test cases should be part of TestTorch similar to what we've done for MPI here: https://github.com/kubeflow/trainer/pull/2481

andreyvelich reviewed (commented):
Thank you for working on this @willb!

Diasker commented:
@andreyvelich Thank you for your feedback! I've addressed all your comments:

1. Regarding the SDK comment: I've removed the comment.

2. For CPU support in torch.go: I've implemented the logic to handle `numProcPerNode="cpu"` which will always use CPU resources regardless of GPU presence. This ensures users can explicitly choose to use CPU cores for distributed training even when GPUs are available.

3. Following your suggestion and PR #2481, I've created a comprehensive `torch_test.go` file instead of a separate test file. The test cases cover various scenarios:
   - `numProcPerNode=auto` with CPU limits but no GPU
   - `numProcPerNode=auto` with GPU resources
   - `numProcPerNode=cpu` with only CPU resources
   - `numProcPerNode=cpu` with both CPU and GPU resources
   - Handling of fractional CPU values (rounding up)
   - Multi-node multi-GPU training configurations

Is the current CPU handling logic in torch.go robust enough? I've ensured it checks both limits and requests, handles fractional values by rounding up, and guarantees at least one process.

Also, regarding the SDK's `get_num_proc_per_node` function - should we update it to match the Go implementation, particularly to support the "cpu" option? The current SDK function prioritizes GPU resources and doesn't have special handling for the "cpu" value.

Regarding the helper function to check for resource existence: I couldn't find a specific utility function in k8s.io/utils for checking whether a resource exists in container.resources. If such a helper exists, I'd be happy to simplify the code by using it.


As a coding newcomer, I really appreciate your patience and guidance throughout this process. Your feedback has been invaluable in helping me improve the implementation.

Diasker commented:
/reset

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch_test.go:
```diff
@@ -0,0 +1,516 @@
+				Scheduler: &runtime.Scheduler{TotalRequests: map[string]runtime.TotalResourceRequest{}},
+			},
+		},
+		// Issue #2407 test case - nproc_per_node=auto with CPU limit
```
```suggestion
		// nproc_per_node=auto with CPU limit
```

It would be better not to add such a specific prefix.
Could you address those in other places, as well?

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch_test.go:
```diff
@@ -0,0 +1,516 @@
+				t.Errorf("Unexpected error from EnforceMLPolicy (-want,+got):\n%s", diff)
+			}
+
+			// If need to validate numProcPerNode
+			if tc.wantNumProcPerNode != "" && tc.info != nil {
+				// Find PET_NPROC_PER_NODE environment variable
+				var numProcPerNodeValue string
+				for _, env := range tc.info.Trainer.Env {
+					if env.Name != nil && *env.Name == constants.TorchEnvNumProcPerNode {
+						if env.Value != nil {
+							numProcPerNodeValue = *env.Value
+						}
+						break
+					}
+				}
+
+				if diff := cmp.Diff(tc.wantNumProcPerNode, numProcPerNodeValue); diff != "" {
+					t.Errorf("Torch.EnforceMLPolicy() numProcPerNode mismatch (-want +got):\n%s", diff)
+				}
+			}
```
We should always compare whole of expected and got runtime.Info.

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +75,78 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
 	}
 
+	// Cap nproc_per_node based on CPU resources when set to "auto" and no GPU is requested
+	if numProcPerNode.String() == "auto" && trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.ResourcesPerNode != nil {
```
These implementation looks slightly redundant. We can do something like

```go
func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) error {

    [...]

	if jobTrainer := trainJob.Spec.Trainer; jobTrainer != nil && jobTrainer.ResourcesPerNode != nil {
		var (
			shouldUseCPU           func(resources corev1.ResourceList) bool
			fallbackNumProcPerNode intstr.IntOrString
		)
		switch numProcPerNode.String() {
		case "auto":
			shouldUseCPU = func(resources corev1.ResourceList) bool {
				for resName := range resources {
					if strings.Contains(strings.ToLower(resName.String()), "gpu") { return false }
				}
				return true
			}
			fallbackNumProcPerNode = intstr.FromString("auto")
		case "cpu":
			shouldUseCPU = func(resources corev1.ResourceList) bool {
				_, ok := resources[corev1.ResourceCPU]
				return ok
			}
			fallbackNumProcPerNode = intstr.FromInt32(1)
		default:
			shouldUseCPU = func(resources corev1.ResourceList) bool { return false }
			fallbackNumProcPerNode = numProcPerNode
		}
		nppNode, usedCPU := calculateNumProcPerNode(fallbackNumProcPerNode, jobTrainer.ResourcesPerNode.Limits, shouldUseCPU)
		if !usedCPU {
			nppNode, usedCPU = calculateNumProcPerNode(fallbackNumProcPerNode, jobTrainer.ResourcesPerNode.Requests, shouldUseCPU)
		}
		numProcPerNode = nppNode
	}

    [...]

}

func calculateNumProcPerNode(
	fallbackNumProcPerNode intstr.IntOrString, resources corev1.ResourceList, shouldUseCPU func(resources corev1.ResourceList) bool,
) (intstr.IntOrString, bool){
	var defaultCPU int32 = 1
	if resources != nil {
		if shouldUseCPU(resources) {
			cpuQ := resources[corev1.ResourceCPU]
			return intstr.FromInt32(max(defaultCPU, int32(cpuQ.Value()))), true
		}
		return fallbackNumProcPerNode, false
	}
	return intstr.FromInt32(defaultCPU), false
}
```

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch_test.go:
```diff
@@ -0,0 +1,1012 @@
+				t.Errorf("Unexpected error from EnforceMLPolicy (-want,+got):\n%s", diff)
+			}
+
+			// Validate the entire info object (if wantInfo is provided)
+			if tc.wantInfo != nil {
```
```suggestion
```

Even if `wantInfo` is nil, we should confirm if got info is null.

Diasker commented:
@tenzen-y  Thanks for your feedback! I've finished change as suggested. As a newcomer, I greatly appreciate your patience and guidance.

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -119,3 +154,30 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+	if resources != nil {
+		if shouldUseCPU(resources) {
+			cpuQ := resources[corev1.ResourceCPU]
+			if !cpuQ.IsZero() {
+				// Round up to whole cores
+				cpuQ.RoundUp(0) // Scale 0 means whole units (cores)
+
+				// Ensure at least 1 process
+				cpuValue := cpuQ.Value()
+				if cpuValue < int64(defaultCPU) {
+					cpuValue = int64(defaultCPU)
+				}
+
+				return intstr.FromInt(int(cpuValue)), true
+			}
```
```suggestion
			return intstr.FromInt32(max(defaultCPU, cpuQ))
```

Why do we need such verification logic? Isn't this enough?
Additionally, `FromInt` is a deprecated function.

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch_test.go:
```diff
@@ -0,0 +1,1014 @@
+			}
+
+			// Validate the entire info object
+			if tc.wantInfo != nil {
+				if diff := cmp.Diff(tc.wantInfo, tc.info,
+					cmpopts.SortSlices(func(a, b string) bool { return a < b }),
+					cmpopts.SortMaps(func(a, b string) bool { return a < b }),
+				); len(diff) != 0 {
+					t.Errorf("Unexpected RuntimeInfo (-want,+got):\n%s", diff)
+				}
+			} else if tc.info != nil {
+				t.Errorf("Expected info to be nil, but got: %v", tc.info)
+			}
```
```suggestion
			if diff := cmp.Diff(tc.wantInfo, tc.info,
				cmpopts.SortSlices(func(a, b string) bool { return a < b }),
				cmpopts.SortMaps(func(a, b string) bool { return a < b }),
			); len(diff) != 0 {
				t.Errorf("Unexpected RuntimeInfo (-want,+got):\n%s", diff)
			}
```

Why can't we do this one?

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch_test.go:
```diff
@@ -0,0 +1,1014 @@
+				Scheduler: &runtime.Scheduler{TotalRequests: map[string]runtime.TotalResourceRequest{}},
+			},
+		},
+		// nproc_per_node=auto with CPU limit
```
We have a description in the map key. So we can remove such comments.

Diasker replied:
@tenzen-y  Thank you for your suggestion to simplify the code. You're right - the max function already handles the case when cpuQ is zero, making the cpuQ.IsZero() check unnecessary. I apologize for the overly complex logic in my previous implementation. Your approach is much cleaner while maintaining the same functionality. I've updated the code accordingly.

tenzen-y reviewed (commented):
Thank you for update. I left a few comments for just nits.
Additionally, could you share the code test coverage for the torch plugin (`pkg/runtime/framework/plugins/torch`)?

Diasker commented:
@tenzen-y  Thank you for your comments. I've addressed the nits you mentioned.
Regarding the code coverage for the torch plugin, here are the results:
```
github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:46:    New                     100.0%
github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:50:    Name                    0.0%
github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:55:    Validate                0.0%
github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:60:    EnforceMLPolicy         94.6%
github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:160:   calculateNumProcPerNode 100.0%
total:                                                                          (statements)            91.5%
```
Is necessary for us to achieve 100% coverage for EnforceMLPolicy function?

tenzen-y commented:
> github.com/kubeflow/trainer/pkg/runtime/framework/plugins/torch/torch.go:55:    Validate                0.0%

I do not offer 100% coverage. It looks like enough. Thank you!

astefanutti commented on sdk/kubeflow/trainer/utils/utils.py:
```diff
@@ -111,8 +111,6 @@ def get_resources_per_node(
 
-# TODO (andreyvelich): Move this part to the Kubeflow Trainer torch plugins.
-# Ref issue: https://github.com/kubeflow/trainer/issues/2407
 def get_num_proc_per_node(resources_per_node: dict) -> str:
```
I think this can be removed altogether.

Diasker replied:
Right, i'll remove the function later.

tenzen-y commented:
@Diasker It looks like there are conflicts

tenzen-y commented:
Part of https://github.com/kubeflow/trainer/issues/2468

astefanutti commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -119,3 +154,19 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+		}
+		return fallbackNumProcPerNode, false
+	}
+	return intstr.FromInt32(int32(defaultCPU)), false
```
is the conversion to int32 really needed?

Diasker replied:
@astefanutti Yes, the conversion is necessary because FromInt32 requires an int32 parameter.
Or we can define ```defaultCPU``` as int32 from the beginning: ```var defaultCPU int32 = 1```

astefanutti replied:
Isn't it already declared as `var defaultCPU int32 = 1`?

Diasker replied:
@astefanutti OMG!!! You're absolutely right! It is already declared as `var defaultCPU int32 = 1`. I apologize for the confusion - I must have overlooked this when reviewing the code. Thank you for pointing this out.

Diasker commented:
> @Diasker It looks like there are conflicts

@tenzen-y Yes, I see the conflicts. This is similar to the issue we discussed earlier with the trainingruntime_test.go file.

> 
> I noticed a test failure in `pkg/runtime/core/trainingruntime_test.go` in the test case `succeeded to build JobSet with Torch values from the Runtime and envs` within `TestTrainingRuntimeNewObjects`.
> 
> The current implementation logic for `nproc_per_node` is:
> 
> * When set to "auto" with no GPU resources requested: use CPU resource limits to set the process count
> * When GPU resources are requested: keep "auto" value
> 
> The existing test case expects `nproc_per_node` to remain "auto", but since it only defines CPU resources (without GPU), the test is failing. Would you like me to change this test file? Maybe we can:
> 
> 1. Update the test case to include GPU resources to maintain the "auto" value, or
> 2. Update the test's expected value to "1" to match the CPU-based calculation?
> 

I'd like to use the same approach for trainjob_controller_test.go as we did for trainingruntime_test.go, but I'm encountering a merge conflict.

tenzen-y commented:
> I'd like to use the same approach for trainjob_controller_test.go as we did for trainingruntime_test.go

@Diasker yes, you can take the same approach. Is there any blocker  for git rebase?

Diasker commented:
@tenzen-y No blockers for git rebase. I was just unfamiliar with how to resolve this specific conflict, but I've figured it out now. Thanks!

andreyvelich commented on sdk/kubeflow/trainer/utils/utils.py:
```diff
@@ -101,31 +101,6 @@ def get_resources_per_node(
 
-# TODO (andreyvelich): Move this part to the Kubeflow Trainer torch plugins.
-# Ref issue: https://github.com/kubeflow/trainer/issues/2407
-def get_num_proc_per_node(resources_per_node: dict) -> str:
```
This you removed this func, you also need to remove it from here: https://github.com/kubeflow/trainer/blob/b3eabd33d1e3ce2ebf5d028fb6113c0921466199/sdk/kubeflow/trainer/api/trainer_client.py#L225-L230

Diasker replied:
@andreyvelich  Alright, thanks!

tenzen-y commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +74,39 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+			}
+			fallbackNumProcPerNode = intstr.FromString("auto")
+		case "cpu":
+			shouldUseCPU = func(resources corev1.ResourceList) bool {
```
```suggestion
			shouldUseCPU = func(corev1.ResourceList) bool {
```

nits: This is not used anywhere.

Diasker replied:
Sure, I'll open another PR to address it later. Thanks for your Approval!

tenzen-y reviewed (commented):
Thank you for this great contribution!
I lest just a nit comment. Could you open another PR to address it?

/lgtm
/approve

andreyvelich commented on pkg/runtime/framework/plugins/torch/torch.go:
```diff
@@ -72,6 +74,39 @@ func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob)
+		case "auto":
+			shouldUseCPU = func(resources corev1.ResourceList) bool {
+				for resName := range resources {
+					if strings.Contains(strings.ToLower(resName.String()), "gpu") {
+						return false
```
This might not work for other devices (e.g. `tpu`, `npu`).
@tenzen-y @Diasker @astefanutti Should we exclude all of them ?

tenzen-y replied:
I think it's challenging to cover all corner case since we can specify arbitrary name for special device in resources fields. 

So, I think it would be better to add device names based on user feedbacks. I meant I want to find common sense for actual used special device names based on feedbacks.

Additionally, there are limitLange problem and etc.

Diasker replied:
I agree with @tenzen-y's approach. Trying to cover all possible device types proactively would be challenging since resource names can be arbitrary. It makes more sense to start with GPU support (which is the most common case) and then extend to other accelerators like TPU/NPU based on actual user feedback and real-world usage patterns. If users report issues with specific devices, we can add support for those devices in future iterations.

astefanutti replied:
+1

andreyvelich replied:
sounds good.

astefanutti commented:
@Diasker thanks for the fantastic work!

Diasker commented:
> @Diasker thanks for the fantastic work!
@astefanutti My pleasure. Thanks for your guidance and patience.
